### PR TITLE
Backend: register

### PR DESCRIPTION
**Pull request by t-h2o**
# /register

inspired by this [example](https://github.com/pallets/flask/blob/2.2.5/examples/tutorial/flaskr/auth.py)

## Curl

```bash
curl localhost:5001/register \
  --data "username=user" \
  --data "firstname=firstname" \
  --data "lastname=lastname" \
  --data "email=email@email.com" \
  --data "password=1234"
```

## PostgreSQL

Look in table in database

```bash
docker exec postgres psql \
  $(grep DATABASE_URL= .env | sed 's/DATABASE_URL=//') \
  -c "select * from users;"
```

### CODE REVIEW

File: backend/app.py [hunk 1]
```diff
@@ -16,20 +18,115 @@ def hello_world():
...
+                error = f"User {username} is already registered."
+            else:
+                return f"User {username} was succefull added"
+            finally:
+                conn.close()
+
+    return f"error: {error}"
+
+
+@app.route("/drop", methods=("GET", "POST"))
```
Comment (RobinBurri) [inline backend/app.py:102]:
```suggestion
@app.route("/drop", methods=["POST"])
```

Comment (RobinBurri) [reply backend/app.py:102]: same than before

File: backend/app.py [hunk 2]
```diff
@@ -16,20 +18,115 @@ def hello_world():
...
+def drop_table():
+    if request.method == "GET":
+        return "drop with POST method"
+    elif request.method == "POST":
```
Comment (RobinBurri) [inline backend/app.py:106]:
```suggestion
    if request.method == "GET":
        return "drop with POST method"
    elif request.method == "POST":
```

Comment (RobinBurri) [reply backend/app.py:106]: same than before

Attention indentation sera fausse pour la suite du code

File: backend/app.py [hunk 3]
```diff
@@ -16,20 +18,115 @@ def hello_world():
...
+        error = None
+
+        if not table:
+            error = "table is required."
+
+        conn = get_db_connection()
+
+        if error is None:
+            try:
+                cur = conn.cursor()
```
Comment (RobinBurri) [inline backend/app.py:118]:
```suggestion
                with conn.cursor() as cur:
```

Comment (RobinBurri) [reply backend/app.py:118]: same

File: backend/app.py [hunk 4]
```diff
@@ -16,20 +18,115 @@ def hello_world():
...
+                cur.execute(f"DROP table IF EXISTS {table}")
+                conn.commit()
+            except conn.IntegrityError:
+                error = f"Table doesn't exist."
+            else:
+                return f'Table "{table}" was succefull dropped'
+            finally:
+                conn.close()
```
Comment (RobinBurri) [inline backend/app.py:126]:
```suggestion
```

Comment (RobinBurri) [reply backend/app.py:126]: same

**Review by RobinBurri — APPROVED** [review]
👍

File: backend/app.py [hunk 5]
```diff
@@ -16,20 +18,115 @@ def hello_world():
...
+    """
+
+    if request.method == "GET":
+        return "register with POST method"
+    elif request.method == "POST":
+        username = request.form["username"]
+        password = request.form["password"]
+        firstname = request.form["firstname"]
+        lastname = request.form["lastname"]
+        email = request.form["email"]
```
[the remark this answers is not in the document]
Comment (t-h2o) [reply backend/app.py]: Done: 92346f7090cfddcf33f2ab3c46384c1b454572b3

File: backend/app.py (hunk 2 above)
Comment (t-h2o) [reply backend/app.py]: done: c9791cc5452f4d1ff1c3a4adfb6b08272acdf4f0

File: backend/app.py (hunk 3 above)
Comment (t-h2o) [reply backend/app.py:112]: Done: 5087c3a4ef11ae4996c562b261c3a3e43d05ca81

File: backend/app.py [hunk 6]
```diff
@@ -16,20 +18,115 @@ def hello_world():
...
+                        email,
+                    ),
+                )
+                conn.commit()
+            except conn.IntegrityError:
+                error = f"User {username} is already registered."
+            else:
+                return f"User {username} was succefull added"
+            finally:
+                conn.close()
```
[the remark this answers is not in the document]
Comment (t-h2o) [reply backend/app.py]: Done 5087c3a4ef11ae4996c562b261c3a3e43d05ca81

File: backend/app.py (hunk 4 above)
Comment (t-h2o) [reply backend/app.py]: done: 5087c3a4ef11ae4996c562b261c3a3e43d05ca81

File: backend/app.py [hunk 7]
```diff
@@ -1,34 +1,135 @@
...
+
+    with get_db_connection() as conn:
+        if error is None:
+            try:
+                cur = conn.cursor()
+                cur.execute(f"DROP table IF EXISTS {table}")
+                conn.commit()
+                return f'Table "{table}" was succefull dropped'
+            except:
+                error = f"drop database exception"
```
Comment (RobinBurri) [inline backend/app.py:130]:
```suggestion
                error = "drop database exception"
```
simple string is enough

**Review by RobinBurri — APPROVED** [review]
check indentation si tu accept mes proposition.
👍

File: backend/app.py [hunk 8]
```diff
@@ -1,34 +1,135 @@
...
+    if not username:
+        error = "Username is required."
+    elif not password:
+        error = "Password is required."
+    elif not firstname:
+        error = "Firstname is required."
+    elif not lastname:
+        error = "Lastname is required."
+    elif not email:
+        error = "Email is required."
```
[the remark this answers is not in the document]
Comment (t-h2o) [reply backend/app.py:81]: J'ai rajouté quelques deux checks dessus : dd66fd27769e9aa5e8d9152a9ae6d3451b6fc4a8

[the remark this answers is not in the document]
Comment (RobinBurri) [reply backend/app.py:81]: Table users {
  id integer [primary key, unique]
  username varchar [unique, not null]
  password varchar [not null]
  firstname varchar [not null]
  lastname varchar [not null]
  email varchar [unique, not null]
  created_at timestamp
}

la db va raise un exception si une valeur est null car on precise [not null]

**Pull request merged**